File: array/862.cpp
#include<iostream>
#include<vector>
#include<deque>
using namespace std;


int shortestSubarray(vector<int>& A, int K) {
	int len = A.size();
        vector<int> sums(len+1, 0);
        deque<int> queue;
        for (int i=0; i<len; i++)
                sums[i+1] = sums[i] + A[i];
        int res = len + 1;

        for (int i=0; i<len+1; i++) {
                while (!queue.empty() && sums[i]<=sums[queue.back()]){
                        queue.pop_back();
                }
                while (!queue.empty() && sums[i]-sums[queue.front()]>=K){
                        res = min(res, i-queue.front());
                        queue.pop_front();
                }
                queue.push_back(i);
        }
        if (res==len+1) res = -1;
        return res;
}

int main() {
	vector<int> A {2, -1, 2};
	
	int K = 3;
	int min_len = shortestSubarray(A, K);
	std::cout<<min_len;
}

File: array/862.py
"""
https://leetcode-cn.com/problems/shortest-subarray-with-sum-at-least-k/
"""

from typing import List

class Solution:
    def shortestSubarray(self, A: List[int], K: int) -> int:
        sums = [0] * (len(A)+1)
        sums[0] = 0

        for i in range(0, len(A)): sums[i+1] = sums[i] + A[i]

        queue = []
        res = len(A) + 1
        for i, p in enumerate(sums):
            while len(queue)>0 and p<=sums[queue[-1]]:
                queue = queue[:-1]

            while len(queue)>0 and p-sums[queue[0]]>=K:
                res = min(res, i-queue[0])
                queue = queue[1:]

            queue += [i]
        if res == len(A) + 1: res = -1
        return res

if __name__ == "__main__":
    s = Solution()
    A = [2,-1,2]
    K = 3
    min_len = s.shortestSubarray(A, K)
    print(min_len)


